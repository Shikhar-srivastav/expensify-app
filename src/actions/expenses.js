import { v4 as uuidv4 } from 'uuid';

export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        _id: uuidv4(),
        description,
        note,
        amount,
        createdAt
    }
});

export const removeExpense = (_id) => ({
    type: 'REMOVE_EXPENSE',
    _id
});

export const editExpense = (_id, updates) => ({
    type: 'EDIT_EXPENSE',
    _id,
    updates
});