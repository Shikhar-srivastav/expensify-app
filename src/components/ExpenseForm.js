import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        }
    }
    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = e => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = createdAt => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = e => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please enter valid details!' }));
        } else {
            this.setState(() => ({ error: undefined }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit} className='form'>
                    {!!this.state.error && <p>{this.state.error}</p>}
                    <div className='input-group__item'>
                        <input
                            className='text-input'
                            type='text'
                            placeholder='Description'
                            autoFocus
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                    </div>
                    <div className='input-group__item'>
                        <input
                            className='text-input'
                            type='text'
                            placeholder='Amount'
                            value={this.state.amount}
                            onChange={this.onAmountChange}
                        />
                    </div>
                    <div className='input-group__item'>
                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={day => false}
                        />
                    </div>
                    <div className='input-group__item'>
                        <textarea
                            className='text-area'
                            placeholder='Add a note for your expense (optional)'
                            value={this.state.note}
                            onChange={this.onNoteChange}
                        ></textarea>
                    </div>
                    <div>
                        <button className='button'>
                            Save Expense
                        </button>
                    </div>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;