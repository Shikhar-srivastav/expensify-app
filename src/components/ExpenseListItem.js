import React from 'react';
import { Link, resolvePath } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = props => (
    <div>
        <Link to={`/edit/${props._id}`}>
            <h3>{props.description}</h3>
        </Link>
        <p>
            {`Rs${numeral(props.amount).format('0,0.00')} - ${moment(props.createdAt).format('Do MMMM YYYY')}`}
        </p>
    </div>
);

export default ExpenseListItem;