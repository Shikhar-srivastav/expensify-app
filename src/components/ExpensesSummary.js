import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getFilteredExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/total';

export const ExpensesSummary = props => (
    <div>
        {`Viewing ${props.total.count} expense(s) totalling Rs${numeral(props.total.amount).format('0,0.00')}`}
     </div>
);

const mapStatetoProps = (state) => ({
    total: getTotalExpenses(getFilteredExpenses(state.expenses, state.filters)),
});

export default connect(mapStatetoProps)(ExpensesSummary)