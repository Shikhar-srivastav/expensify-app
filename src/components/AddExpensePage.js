import React from "react";
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

export class AddExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.startAddExpense(expense);
        this.props.navigate('/dashboard');
    }
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='container'>
                        <h1 className='page-title'>Add Expense</h1>
                    </div>
                </div>
                <div className='container'>
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ({
    navigate: useNavigate(),
    startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);