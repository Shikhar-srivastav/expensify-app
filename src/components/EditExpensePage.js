import React from "react";
import { connect } from 'react-redux';
import { useParams, useNavigate } from "react-router-dom";
import ExpenseForm from '../components/ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.editExpense(this.props.params._id, expense);
        this.props.navigate('/');
    }
    onClick = () => {
        this.props.removeExpense(this.props.params._id);
        this.props.navigate('/');
    }
    currentExpense = this.props.expenses.find(({ _id }) => (
        _id === this.props.params._id
    ));
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm
                    expense={this.currentExpense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onClick}
                >Remove</button>
            </div>
        );
    }
}



const mapDispatchToProps = (dispatch) => ({
    editExpense: (_id, expense) => {
        dispatch(editExpense(_id, expense));
    },
    removeExpense: (_id) => {
        dispatch(removeExpense(_id));
    },
    navigate: useNavigate(),
    params: useParams()
});

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);