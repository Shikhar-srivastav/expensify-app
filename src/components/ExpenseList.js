import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getFilteredExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map(expense => (
                    <ExpenseListItem
                        key={expense._id}
                        {...expense}
                    />
                ))
            )
        }
    </div>
);

const ConnectedExpenseList = connect(state => {
    return {
        expenses: getFilteredExpenses(state.expenses, state.filters)
    };
})(ExpenseList);

export default ConnectedExpenseList;