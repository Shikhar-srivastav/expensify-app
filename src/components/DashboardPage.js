import React from "react";
import ExpenseList from './ExpenseList';
import ExpenseFilters from './ExpenseFilters';

const DashboardPage = () => (
    <div>
        <ExpenseFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;