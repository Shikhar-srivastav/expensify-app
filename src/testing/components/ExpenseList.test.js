import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('Render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('Render ExpenseList with empty list', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});

