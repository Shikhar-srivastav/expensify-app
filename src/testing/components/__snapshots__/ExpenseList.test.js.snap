// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExpenseList with empty list 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`Render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    _id="1"
    amount={10}
    createdAt={0}
    description="Gum"
    key="1"
    note=""
  />
  <ExpenseListItem
    _id="2"
    amount={3000}
    createdAt={-345600000}
    description="Credit Card"
    key="2"
    note=""
  />
  <ExpenseListItem
    _id="3"
    amount={10500}
    createdAt={345600000}
    description="Rent"
    key="3"
    note=""
  />
</div>
`;
